subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1634 @@
+	hash = id;
+	hash ^= hash >> 16;
+	hash *= 0x7feb352dU;
+	hash ^= hash >> 15;
+	hash *= 0x846ca68bU;
+	hash ^= hash >> 16;
+
+	return hash % CONFIG_ZMS_LOOKUP_CACHE_SIZE;
+}
+
+static int zms_lookup_cache_rebuild(struct zms_fs *fs)
+{
+	int rc, previous_sector_num = -1;
+	uint64_t addr, ate_addr;
+	uint64_t *cache_entry;
+	uint8_t current_cycle;
+	struct zms_ate ate;
+
+	memset(fs->lookup_cache, 0xff, sizeof(fs->lookup_cache));
+	addr = fs->ate_wra;
+
+	while (true) {
+		/* Make a copy of 'addr' as it will be advanced by zms_prev_ate() */
+		ate_addr = addr;
+		rc = zms_prev_ate(fs, &addr, &ate);
+
+		if (rc) {
+			return rc;
+		}
+
+		cache_entry = &fs->lookup_cache[zms_lookup_cache_pos(ate.id)];
+
+		if (ate.id != ZMS_HEAD_ID && *cache_entry == ZMS_LOOKUP_CACHE_NO_ADDR) {
+			/* read the ate cycle only when we change the sector
+			 * or if it is the first read
+			 */
+			if ((SECTOR_NUM(ate_addr) != previous_sector_num) ||
+			    (previous_sector_num == -1)) {
+				rc = zms_get_sector_cycle(fs, ate_addr, &current_cycle);
+				if (rc) {
+					return rc;
+				}
+			}
+			if (zms_ate_valid_different_sector(fs, &ate, current_cycle)) {
+				*cache_entry = ate_addr;
+			}
+			previous_sector_num = SECTOR_NUM(ate_addr);
+		}
+
+		if (addr == fs->ate_wra) {
+			break;
+		}
+	}
+
+	return 0;
+}
+
+static void zms_lookup_cache_invalidate(struct zms_fs *fs, uint32_t sector)
+{
+	uint64_t *cache_entry = fs->lookup_cache;
+	uint64_t *const cache_end = &fs->lookup_cache[CONFIG_ZMS_LOOKUP_CACHE_SIZE];
+
+	for (; cache_entry < cache_end; ++cache_entry) {
+		if ((*cache_entry >> ADDR_SECT_SHIFT) == sector) {
+			*cache_entry = ZMS_LOOKUP_CACHE_NO_ADDR;
+		}
+	}
+}
+
+#endif /* CONFIG_ZMS_LOOKUP_CACHE */
+
+/* zms_al_size returns size aligned to fs->write_block_size */
+static inline size_t zms_al_size(struct zms_fs *fs, size_t len)
+{
+	size_t write_block_size = fs->flash_parameters->write_block_size;
+
+	if (write_block_size <= 1U) {
+		return len;
+	}
+	return (len + (write_block_size - 1U)) & ~(write_block_size - 1U);
+}
+
+/* Aligned memory write */
+static int zms_flash_al_wrt(struct zms_fs *fs, uint64_t addr, const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc = 0;
+	off_t offset;
+	size_t blen;
+	uint8_t buf[ZMS_BLOCK_SIZE];
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	if (!len) {
+		/* Nothing to write, avoid changing the flash protection */
+		return 0;
+	}
+
+	offset = fs->offset;
+	offset += fs->sector_size * (addr >> ADDR_SECT_SHIFT);
+	offset += addr & ADDR_OFFS_MASK;
+
+	blen = len & ~(fs->flash_parameters->write_block_size - 1U);
+	if (blen > 0) {
+		rc = flash_write(fs->flash_device, offset, data8, blen);
+		if (rc) {
+			/* flash write error */
+			goto end;
+		}
+		len -= blen;
+		offset += blen;
+		data8 += blen;
+	}
+	if (len) {
+		memcpy(buf, data8, len);
+		if (ebw_required) {
+			(void)memset(buf + len, fs->flash_parameters->erase_value,
```

Comment:
Why is this done only for `eraseable` devices? It's better to fill the remaining space with known values than to write an uninitialized stack buffer data.